Log indexer panic: send on closed channel during early runServe exit

## Bug

When `runServe` returns an error early in startup (before the HTTP server binds), the log indexer goroutine panics with `send on closed channel` at `internal/logging/indexer.go:169`.

## Reproduction

Configure an invalid `provenance.signing_key` path (e.g., bare filename without `~/` prefix). The provenance init fails, `runServe` returns, launchd restarts the process, and the crash-loop produces repeated panics:

```
load provenance signing key id_ed25519_aimee: provenance: read signing key id_ed25519_aimee: open id_ed25519_aimee: no such file or directory
panic: send on closed channel

goroutine 20 [running]:
github.com/nugget/thane-ai-agent/internal/logging.(*IndexHandler).Handle(...)
	github.com/nugget/thane-ai-agent/internal/logging/indexer.go:169 +0x4b4
log/slog.(*Logger).Debug(...)
main.runServe.func1()
	github.com/nugget/thane-ai-agent/cmd/thane/main.go:415 +0x214
```

## Root Cause

The `IndexHandler` channel is closed during shutdown, but a background goroutine (`runServe.func1` at line 415 — the log index pruner) still attempts to log via `slog.Debug` after the channel is closed. The goroutine outlives the handler teardown.

## Fix

Either:
1. Guard the channel send with a select + done channel (or recover from panic)
2. Ensure the pruner goroutine is cancelled before closing the indexer channel
3. Use a sync mechanism (e.g., `sync.Once` on close) so sends to a closed channel are no-ops

Option 1 is simplest — a `select { case ch <- record: default: }` or recover guard in `Handle()` prevents the panic without requiring goroutine lifecycle changes.

## Impact

Any early startup failure (bad config, missing files, network issues) that causes `runServe` to return before full initialization can trigger this panic. The panic itself is harmless (process was already dying) but it pollutes stderr and masks the real error in crash-loop scenarios.

## Version

v0.8.3-50-gace98fc

--- a/internal/logging/indexer.go
+++ b/internal/logging/indexer.go
@@ -10,7 +10,6 @@ import (
 	"slices"
 	"strings"
 	"sync"
-	"sync/atomic"
 	"time"
 )
 
@@ -59,8 +58,9 @@ type indexShared struct {
 	db     *sql.DB
 	ch     chan indexEntry
 	done   chan struct{}
-	closed atomic.Bool // true after Close; guards sends on ch
-	once   sync.Once   // guards Close
+	mu     sync.RWMutex // serializes Handle sends with Close
+	closed bool         // true after Close; protected by mu
+	once   sync.Once    // guards Close
 }
 
 // indexEntry is the set of fields written to SQLite for one log record.
@@ -166,14 +166,19 @@ func (h *IndexHandler) Handle(ctx context.Context, r slog.Record) error {
 		entry.Attrs = string(b)
 	}
 
-	// Drop the entry if the handler has been closed or the channel is
-	// full. Never block a log call on database I/O.
-	if !h.shared.closed.Load() {
+	// Non-blocking send under a read lock so Close() cannot close the
+	// channel while we're sending. The closed flag is checked inside
+	// the lock to eliminate the TOCTOU race between checking and
+	// sending. Drop the entry if the channel is full — never block a
+	// log call on database I/O.
+	h.shared.mu.RLock()
+	if !h.shared.closed {
 		select {
 		case h.shared.ch <- entry:
 		default:
 		}
 	}
+	h.shared.mu.RUnlock()
 
 	return nil
 }
@@ -207,8 +212,10 @@ func (h *IndexHandler) WithGroup(name string) slog.Handler {
 // goroutine. It is safe to call multiple times.
 func (h *IndexHandler) Close() {
 	h.shared.once.Do(func() {
-		h.shared.closed.Store(true)
+		h.shared.mu.Lock()
+		h.shared.closed = true
 		close(h.shared.ch)
+		h.shared.mu.Unlock()
 		<-h.shared.done
 	})
 }

--- a/internal/logging/indexer_test.go
+++ b/internal/logging/indexer_test.go
@@ -360,6 +360,33 @@ func TestIndexHandler_LogAfterCloseNoPanic(t *testing.T) {
 	logger.Debug("another post-close log")
 }
 
+func TestIndexHandler_ConcurrentLogDuringClose(t *testing.T) {
+	db := openTestDB(t)
+	inner := slog.NewJSONHandler(discardWriter{}, nil)
+	h := NewIndexHandler(inner, db, nil)
+
+	logger := slog.New(h)
+
+	// Simulate the real crash scenario: a background goroutine
+	// continues logging while Close() runs concurrently.
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for range 1000 {
+			logger.Info("background log")
+		}
+	}()
+
+	// Close while the goroutine is still logging.
+	h.Close()
+
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
+		t.Fatal("timed out waiting for background logger")
+	}
+}
+
 func TestPrune_DebugAndTrace(t *testing.T) {
 	db := openTestDB(t)
 

